src/OrleansServiceBus/Providers/Streams/EventHub/SegmentBuilder.cs

```diff
@@ -0,0 +1,135 @@
+﻿using System;
+
+namespace Orleans.ServiceBus.Providers
+{
+    internal static class SegmentBuilder
+    {
+        /// <summary>
+        /// Calculates how much space will be needed to append the provided bytes into the segment.
+        /// </summary>
+        /// <param name="bytes"></param>
+        /// <returns></returns>
+        public static int CalculateAppendSize(byte[] bytes)
+        {
+            return (bytes == null || bytes.Length == 0)
+                ? sizeof(int)
+                : bytes.Length + sizeof(int);
+        }
+
+        /// <summary>
+        /// Calculates how much space will be needed to append the provided string into the segment.
+        /// </summary>
+        /// <param name="str"></param>
+        /// <returns></returns>
+        public static int CalculateAppendSize(string str)
+        {
+            return (string.IsNullOrEmpty(str))
+                ? sizeof(int)
+                : str.Length * sizeof(char) + sizeof(int);
+        }
+
+        /// <summary>
+        /// Appends an array of bytes to the end of the segment
+        /// </summary>
+        /// <param name="writerOffset"></param>
+        /// <param name="bytes"></param>
+        /// <param name="segment"></param>
+        public static void Append(ArraySegment<byte> segment, ref int writerOffset, byte[] bytes)
+        {
+            if (segment.Array == null)
+            {
+                throw new ArgumentNullException("segment");
+            }
+            if (bytes == null)
+            {
+                throw new ArgumentNullException("bytes");
+            }
+
+            if (bytes.Length == 0)
+            {
+                Buffer.BlockCopy(BitConverter.GetBytes(0), 0, segment.Array, segment.Offset + writerOffset, sizeof(int));
```

Comment:
I'll keep this in mind when I do perf runs on this.  If it's a measurable perf gain, it could be worth it, but I don't want to introduce unsafe code unless I've demonstrated advantage.  Good info none-the-less.